### PR TITLE
Gather the reflog entry content tests

### PR DESCRIPTION
**Pull request by tiennou**
This is is splitted from #4294 with just the test case wrangling.

### CODE REVIEW

**Comment by tiennou** [comment]
@pks-t I took care of your review comments, apart from the header thing : AFAICT we don't have headers on test files.

**Comment by pks-t** [comment]
Two issues:

- there's one use of C99-style comments in "tests/refs/reflog/messages.c", which we usually avoid
- there's a declaration after code in `reflog_print`

Looks fine otherwise, thanks!

**Comment by tiennou** [comment]
Good to go ! Travis tripped up on proxy things. Again 😉.

File: tests/refs/reflog/reflog_helpers.c [hunk 1]
```diff
@@ -15,51 +15,107 @@ size_t reflog_entrycount(git_repository *repo, const char *name)
...
+		}
+		if (entry_msg && strcasecmp(message, entry_msg) != 0) {
+			git_buf_printf(&result, "\tMessage: \"%s\" != \"%s\"\n", message, entry_msg);
+		}
+	}
+	if (git_buf_len(&result) != 0) {
+		clar__fail(file, line, "Reflog entry mismatch", git_buf_cstr(&result), 1);
 	}
 
 	git_reflog_free(log);
```
Comment (pks-t) [inline tests/refs/reflog/reflog_helpers.c]: `git_buf_free` missing here for the `result` buffer

File: tests/refs/reflog/reflog_helpers.c [hunk 2]
```diff
@@ -15,51 +15,107 @@ size_t reflog_entrycount(git_repository *repo, const char *name)
...
 		git_object *obj = NULL;
 		if (git_revparse_single(&obj, repo, new_spec) == GIT_OK) {
-			cl_assert_equal_oid(git_object_id(obj), git_reflog_entry_id_new(entry));
+			if (git_oid_cmp(git_object_id(obj), git_reflog_entry_id_new(entry)) != 0) {
+				char err[] = "\"........................................\" != \"........................................\"";
+
+				git_oid_fmt(&err[1], git_object_id(obj));
+				git_oid_fmt(&err[47], git_reflog_entry_id_new(entry));
+
+				git_buf_printf(&result, "\tNew OID: %s\n", err);
```
Comment (pks-t) [inline tests/refs/reflog/reflog_helpers.c]: Huh, this seems a bit fragile. How about

```
git_oid__writebuf(&result, "\tNew OID: \"", git_object_id(obj));
git_oid__writebuf(&result, "\" != \", git_reflog_entry_id_new(entry));
git_buf_puts(&result, '"\n');
```

This would save us from the hardcoded indices. Same applies below.

File: tests/refs/reflog/reflog_helpers.c [hunk 3]
```diff
@@ -15,51 +15,107 @@ size_t reflog_entrycount(git_repository *repo, const char *name)
...
+		if (entry_msg == NULL) {
+			entry_msg = "";
+		}
+		if (entry_msg && strcasecmp(message, entry_msg) != 0) {
+			git_buf_printf(&result, "\tMessage: \"%s\" != \"%s\"\n", message, entry_msg);
+		}
+	}
+	if (git_buf_len(&result) != 0) {
+		clar__fail(file, line, "Reflog entry mismatch", git_buf_cstr(&result), 1);
 	}
```
Comment (pks-t) [inline tests/refs/reflog/reflog_helpers.c]: No need for braces around single-line statements (lines 85 to 98)

File: tests/refs/reflog/reflog_helpers.c [hunk 4]
```diff
@@ -69,14 +125,7 @@ void reflog_print(git_repository *repo, const char *reflog_name)
...
 		const git_reflog_entry *entry = git_reflog_entry_byindex(reflog, idx);
-		git_buf buf = GIT_BUF_INIT;
-		char old_oid[GIT_OID_HEXSZ], new_oid[GIT_OID_HEXSZ];;
-		git_oid_tostr((char *)&old_oid, GIT_OID_HEXSZ, git_reflog_entry_id_old(entry));
-		git_oid_tostr((char *)&new_oid, GIT_OID_HEXSZ, git_reflog_entry_id_new(entry));
-
-		git_buf_printf(&buf, "%ld: %s %s %s %s\n", idx, old_oid, new_oid, "somesig", git_reflog_entry_message(entry));
-		fprintf(stderr, "%s", git_buf_cstr(&buf));
-		git_buf_free(&buf);
+		fprintf(stderr, "%s\n", reflog_entry_tostr(entry));
```
Comment (pks-t) [inline tests/refs/reflog/reflog_helpers.c]: This is a memory leak as `reflog_entry_tostr` returns a string detached from its buffer. I would instead propose an interface `int reflog_entry_tostr(git_buf *out, const git_reflog_entry *entry)`, which makes the need for memory management more obvious.

File: tests/refs/reflog/reflog_helpers.c [hunk 5]
```diff
@@ -15,51 +15,107 @@ size_t reflog_entrycount(git_repository *repo, const char *name)
...
+
+				git_buf_printf(&result, "\tNew OID: %s\n", err);
+			}
 			git__free(oid);
 		}
 	}
 
-	if (email) {
-		cl_assert_equal_s(email, git_reflog_entry_committer(entry)->email);
+	if (email && strcasecmp(email, git_reflog_entry_committer(entry)->email) != 0) {
```
Comment (pks-t) [inline tests/refs/reflog/reflog_helpers.c]: Should this comparison really be case-insensitive?

File: tests/refs/reflog/reflog_helpers.c (hunk 2 above)
Comment (pks-t) [inline tests/refs/reflog/reflog_helpers.c]: And above, as well.

File: tests/refs/reflog/reflog_helpers.c [hunk 6]
```diff
@@ -15,51 +15,107 @@ size_t reflog_entrycount(git_repository *repo, const char *name)
...
 {
 	git_reflog *log;
 	const git_reflog_entry *entry;
+	git_buf result = GIT_BUF_INIT;
 
 	cl_git_pass(git_reflog_read(&log, repo, reflog));
 	entry = git_reflog_entry_byindex(log, idx);
+	if (entry == NULL) {
+		clar__fail(file, line, "Reflog has no such entry", NULL, 1);
+	}
```
Comment (pks-t) [inline tests/refs/reflog/reflog_helpers.c]: No breaces required around single-line statements.

File: tests/refs/reflog/messages.c [hunk 7]
```diff
@@ -37,34 +37,34 @@ void test_refs_reflog_messages__setting_head_updates_reflog(void)
 	cl_git_pass(git_repository_set_head(g_repo, "refs/tags/test"));            /* 1 */
 	cl_git_pass(git_repository_set_head(g_repo, "refs/remotes/test/master"));  /* 0 */
 
-	reflog_check_entry(g_repo, GIT_HEAD_FILE, 4,
```
Comment (pks-t) [inline tests/refs/reflog/messages.c]: I'd prefer if these whitespace changes were put inside another commit, which _only_ changes indentantion. Furthermore, instead of adding more indentation, I'd break with our usual style here and decrease indentation to only one tab on the newline. Otherwise, lines get very long and the statements become quite unreadable.

Comment (tiennou) [inline tests/refs/reflog/messages.c]: Good idea. Folded the indentation change inside the "gathering" step.

File: tests/refs/reflog/reflog_helpers.c (hunk 4 above)
Comment (tiennou) [inline tests/refs/reflog/reflog_helpers.c]: Fixed.

File: tests/refs/reflog/reflog_helpers.c (hunk 1 above)
Comment (tiennou) [inline tests/refs/reflog/reflog_helpers.c]: Fixed

File: tests/refs/reflog/reflog_helpers.c (hunk 5 above)
Comment (tiennou) [inline tests/refs/reflog/reflog_helpers.c]: It should not. Fixed the other occurrence below too.

File: tests/refs/reflog/reflog_helpers.c (hunk 2 above)
Comment (tiennou) [inline tests/refs/reflog/reflog_helpers.c]: Looks better, thanks !

**Comment by tiennou** [comment]
Rebased, squashing review fixups inside their respective commits.

**Comment by tiennou** [comment]
@pks-t Just rebased, ready when you are 😉.

File: tests/refs/reflog/messages.c [hunk 8]
```diff
@@ -0,0 +1,408 @@
...
+void test_refs_reflog_messages__creating_branches_default_messages(void)
+{
+	git_buf buf = GIT_BUF_INIT;
+	git_annotated_commit *annotated;
+	git_object *obj;
+	git_commit *target;
+	git_reference *branch;
+
+	cl_git_pass(git_revparse_single(&obj, g_repo, "e90810b8df3"));
+	cl_git_pass(git_commit_lookup(&target, g_repo, git_object_id(obj)));
```
Comment (pks-t) [inline tests/refs/reflog/messages.c]: The `target` is leaking.

File: tests/refs/reflog/messages.c [hunk 9]
```diff
@@ -0,0 +1,408 @@
...
+	git_object_free(obj);
```
Comment (pks-t) [inline tests/refs/reflog/messages.c]: I'd prefer if you moved all freeing functions to the end. It makes it a lot easier to see if everything has been taken care of by just comparing the list of declarations and free's at the end (at least in most cases).

File: tests/refs/reflog/messages.c [hunk 10]
```diff
@@ -0,0 +1,408 @@
...
+		GIT_OID_HEX_ZERO,
+		git_oid_tostr_s(git_commit_id(target)),
+		g_email, git_buf_cstr(&buf));
+
+	cl_git_pass(git_reference_remove(g_repo, "refs/heads/" NEW_BRANCH_NAME));
+	git_reference_free(branch);
+	git_buf_clear(&buf);
+
+	cl_git_pass(git_annotated_commit_from_revspec(&annotated, g_repo, "e90810b8df3"));
+	cl_git_pass(git_branch_create_from_annotated(&branch, g_repo, NEW_BRANCH_NAME, annotated, true));
```
Comment (pks-t) [inline tests/refs/reflog/messages.c]: `branch` is also leaking

**Comment by tiennou** [comment]
Fixed. I had to append some numbers so there was no variable reused, but I like how clearer it makes those lifetimes.

**Comment by pks-t** [comment]
Thanks a lot for this, @tiennou!
